File: core/Wallet.ts
import { Card } from "./item/Card.ts";
import { Money } from "./item/Money.ts";
import { Sized } from "./item/Sized.ts";
import { Transaction, Transactional } from "./Transaction.ts";

export class Wallet implements Transactional, Sized {
  // in case you don't remember what you did:D
  public readonly actionsHistory: Transaction[];
  private static readonly WALLET_SIZE = 30;
  readonly #cards: Omit<Card, "status">[];
  readonly #balance: Money;

  constructor(public readonly owner: string, initialBalance = 0) {
    this.#balance = Money.from(initialBalance);
    this.#cards = [];
    this.actionsHistory = [];
  }

  deposit(amount: number): Transaction {
    const newDeposit = Money.from(amount);
    this._assertItemFits(newDeposit);
    const transaction = this.#balance.deposit(newDeposit.amount);
    return this._queueHistory(transaction);
  }

  withdraw(amount: number): Transaction {
    const transaction = this.#balance.withdraw(amount);
    return this._queueHistory(transaction);
  }

  putCard(card: Card) {
    this.#cards.push(card);
  }

  getCardByOrder(order: number) {
    const card = this.cards.at(order - 1);
    if (!card) {
      throw new Error("No item at: " + order);
    }
    return card;
  }

  get cards() {
    return this.#cards.slice();
  }

  private _queueHistory(action: Transaction) {
    this.actionsHistory.push(action);
    return action;
  }

  get balance() {
    return this.#balance.amount;
  }

  computeSize(): number {
    return Wallet.WALLET_SIZE;
  }

  public getAvailableSpace() {
    return this.computeSize() - this._computeOccupiedSpace();
  }

  private _assertItemFits(sized: Sized) {
    const available = this.getAvailableSpace();
    if (available < sized.computeSize()) {
      throw new Error(
        "item couldn't fit the wallet anymore, available space is: " +
          available,
      );
    }
  }

  private _computeOccupiedSpace() {
    return [
      this.#balance.computeSize(),
      ...this.cards.map((card) => card.computeSize()),
    ].reduce((agg, curr) => agg + curr, 0);
  }

  toView(): Record<string, unknown> {
    return {
      owner: this.owner,
      balance: this.balance,
      availableSpace: this.getAvailableSpace(),
    };
  }
}

File: core/wallet.py
from core.item.sized import Sized
from core.item.money import Money

WALLET_SIZE = 50

class Wallet(Sized):
    def __init__(self, owner):
        self.owner = owner
        self.balance = Money.from_amount(0)
        self.transaction_history = []
        self.cards = []

    def get_balance(self):
        return self.balance.get_amount()

    def queue_history(self, action):
        self.transaction_history.append(action)
        return action

    def deposit(self, amount):
        money = Money.from_amount(amount)
        self._assert_item_fits(money)
        transaction = self.balance.deposit(money.amount)
        self.transaction_history.append(transaction)
        return transaction

    def withdraw(self, amount):
        try:
            transaction = self.balance.withdraw(amount)
            self.queue_history(transaction)
            return transaction
        except ValueError as e:
            # Handle insufficient funds error
            print(e)
            return None

    def add_card(self, item):
        self._assert_item_fits(item)
        self.cards.append(item)

    def get_card(self, order):
        card = self.cards[order - 1]
        if card == None:
            raise Exception("Card not found")

    def compute_size(self):
        self.balance.compute_size() + map(lambda x: x.compute_size(), self.cards)

    def _assert_item_fits(self, item):
        size = item.compute_size()
        if self._get_available_space() < size:
            raise AssertionError(f"item with size: {size} doesn't fit in the wallet anymore, available_in_wallet={self._get_available_space()}")

    def _get_available_space(self):
        WALLET_SIZE - self.compute_size()